File: system-test/typescript/src/testkit/driver/adapter/external/tax/tax-stub-driver.ts
import type { Result } from '../../../../common/result.js';
import type { TaxErrorResponse } from '../../../port/external/tax/dtos/TaxErrorResponse.js';
import type { ReturnsTaxRateRequest } from '../../../port/external/tax/dtos/ReturnsTaxRateRequest.js';
import { BaseTaxDriver } from './BaseTaxDriver.js';
import { TaxStubClient } from './client/TaxStubClient.js';

export class TaxStubDriver extends BaseTaxDriver<TaxStubClient> {
  constructor(baseUrl: string) {
    super(new TaxStubClient(baseUrl));
  }

  async returnsTaxRate(request: ReturnsTaxRateRequest): Promise<Result<void, TaxErrorResponse>> {
    return this.client.configureTaxRate(request);
  }

  async close(): Promise<void> {
    await this.client.close();
  }
}

File: system-test/java/src/main/java/com/mycompany/myshop/testkit/driver/adapter/external/tax/TaxStubDriver.java
package com.mycompany.myshop.testkit.driver.adapter.external.tax;

import com.mycompany.myshop.testkit.driver.adapter.external.tax.client.TaxStubClient;
import com.mycompany.myshop.testkit.driver.adapter.external.tax.client.dtos.ExtGetCountryResponse;
import com.mycompany.myshop.testkit.driver.port.external.tax.dtos.ReturnsTaxRateRequest;
import com.mycompany.myshop.testkit.driver.port.external.tax.dtos.error.TaxErrorResponse;
import com.mycompany.myshop.testkit.common.Converter;
import com.mycompany.myshop.testkit.common.Result;

public class TaxStubDriver extends BaseTaxDriver<TaxStubClient> {
    public TaxStubDriver(String baseUrl) {
        super(new TaxStubClient(baseUrl));
    }

    @Override
    public void close() {
        client.removeStubs();
        super.close();
    }

    @Override
    public Result<Void, TaxErrorResponse> returnsTaxRate(ReturnsTaxRateRequest request) {
        var country = request.getCountry();
        var taxRate = Converter.toBigDecimal(request.getTaxRate());

        var response = ExtGetCountryResponse.builder()
                .id(country)
                .taxRate(taxRate)
                .countryName(country)
                .build();

        return client.configureGetCountry(country, response)
                .mapError(ext -> new TaxErrorResponse(ext.getMessage()));
    }
}
